File: 0031.NextPermutation/NextPermutation.py
class Solution:
    def nextPermutation(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        nl = len(nums)
        if nl <= 1:
            return nums
        
        loc = 0
        for i in range(1, nl):
            if nums[i] > nums[i - 1]:
                loc = i
        if loc == 0:
            nums = nums.reverse()
            return
        
        mm = loc
        for i in range(loc + 1, nl):
            if nums[i] > nums[loc - 1] and nums[i] < nums[mm]:
                mm = i
        tmp = nums[loc - 1]
        nums[loc - 1] = nums[mm]
        nums[mm] = tmp
        nums[loc:] = sorted(nums[loc:])
        
        return

File: 0031.NextPermutation/NextPermutation.c
int mycompar(const void *a, const void *b)
{
	int *x = (int *)a;
	int *y = (int *)b;
	if (*x > *y) return 1;
	return *x == *y? 0 : -1;
}
void nextPermutation(int* nums, int numsSize) 
{
	int *past = nums, i;
	for (i = 1; i < numsSize; i++) {
		if (nums[i] > nums[i - 1]) past = nums + i;
	}
	if (past == nums) {
		qsort(nums, numsSize, sizeof(int), mycompar);
		return;
	}
	int *left = past - 1;
	int *right = past;
	for (;past - nums < numsSize;past++)
		if (*past < *right && *past > *left) right = past;
	int tmp = *left;
	*left = *right;
	*right = tmp;
	qsort(left + 1, numsSize - (left + 1 - nums), sizeof(int), mycompar);
}